cfbs/core.py

```diff
@@ -0,0 +1,493 @@
+#!/usr/bin/env python3
+import os
+import sys
+import logging as log
+from collections import OrderedDict
+from copy import deepcopy
+
+from cfbs.utils import (
+    cfbs_dir,
+    cfbs_filename,
+    user_error,
+    strip_left,
+    strip_right,
+    pad_right,
+    get_json,
+    write_json,
+    read_json,
+    get_or_read_json,
+    merge_json,
+    read_file,
+    find,
+    mkdir,
+    touch,
+    rm,
+    cp,
+    sh,
+    cfengine_dir,
+)
+from cfbs.internal_file_management import (
+    clone_url_repo,
+    fetch_archive,
+    local_module_name,
+    SUPPORTED_ARCHIVES,
+)
+from cfbs.pretty import pretty
+
+_DEFAULT_INDEX = (
+    "https://raw.githubusercontent.com/cfengine/build-index/master/cfbs.json"
+)
+
+
+def _local_module_data_cf_file(module):
+    target = os.path.basename(module)
+    return {
+        "description": "Local policy file added using cfbs command line",
+        "tags": ["local"],
+        "dependencies": ["autorun"],
+        "steps": ["copy %s services/autorun/%s" % (module, target)],
+        "added_by": "cfbs add",
+    }
+
+
+def _local_module_data_json_file(module):
+    return {
+        "description": "Local augments file added using cfbs command line",
+        "tags": ["local"],
+        "steps": ["json %s def.json" % module],
+        "added_by": "cfbs add",
+    }
+
+
+def _local_module_data_subdir(module):
+    assert module.startswith("./")
+    dst = os.path.join("services", "cfbs", module[2:])
+    return {
+        "description": "Local subdirectory added using cfbs command line",
+        "tags": ["local"],
+        "steps": ["directory {} {}".format(module, dst)],
+        "added_by": "cfbs add",
+    }
+
+
+def _generate_local_module_object(module):
+    assert module.startswith("./")
+    assert module.endswith((".cf", ".json", "/"))
+    assert os.path.isfile(module) or os.path.isdir(module)
+
+    if os.path.isdir(module):
+        return _local_module_data_subdir(module)
+    if module.endswith(".cf"):
+        return _local_module_data_cf_file(module)
+    if module.endswith(".json"):
+        return _local_module_data_json_file(module)
+
+
+class Index:
+    """Class representing the cfbs.json containing the index of available modules"""
+
+    def __init__(self, index=_DEFAULT_INDEX):
+        self._unexpanded = index
+        self._data = None
+
+    def __contains__(self, key):
+        return key in self.data["index"]
+
+    def __getitem__(self, key):
+        return self.data["index"][key]
+
+    def items(self):
+        return self.data["index"].items()
+
+    def get(self, key, default=None):
+        if not key in self:
+            return default
+        return self[key]
+
+    def _expand_index(self):
+        index = self._unexpanded
+        if type(index) in (dict, OrderedDict):
+            self._data = {"type": "index", "index": index}
+            return
+
+        assert type(index) is not Index
+        assert type(index) is str
```

Comment:
`type(index) is str` implies that `type(index) is not Index`, thus  `type(index) is not Index` is excessive

Reply:
Thanks, fixed: https://github.com/cfengine/cfbs/pull/93/commits/f350c7e37d1c3b19d7b2a7b2bb7ebbc2a551f9eb